File: android/app/src/main/kotlin/com/thinkupllc/daystildisney/core/model/DisneyResort.kt
package com.thinkupllc.daystildisney.core.model

/**
 * All six Disney resort destinations worldwide.
 * Each resort holds the list of parks within it and display metadata.
 */
enum class DisneyResort(
    val displayName: String,
    val shortName: String,
    val location: String,
) {
    WALT_DISNEY_WORLD(
        displayName = "Walt Disney World",
        shortName = "WDW",
        location = "Orlando, Florida, USA",
    ),
    DISNEYLAND_RESORT(
        displayName = "Disneyland Resort",
        shortName = "DLR",
        location = "Anaheim, California, USA",
    ),
    TOKYO_DISNEY_RESORT(
        displayName = "Tokyo Disney Resort",
        shortName = "TDR",
        location = "Urayasu, Chiba, Japan",
    ),
    DISNEYLAND_PARIS(
        displayName = "Disneyland Paris",
        shortName = "DLP",
        location = "Marne-la-Vallée, France",
    ),
    HONG_KONG_DISNEYLAND(
        displayName = "Hong Kong Disneyland",
        shortName = "HKDL",
        location = "Lantau Island, Hong Kong",
    ),
    SHANGHAI_DISNEYLAND(
        displayName = "Shanghai Disneyland",
        shortName = "SHDL",
        location = "Pudong, Shanghai, China",
    );

    /** Returns all parks belonging to this resort. */
    val parks: List<DisneyPark>
        get() = DisneyPark.entries.filter { it.resort == this }
}

File: ios/Days-Til-Disney/Days-Til-Disney/Core/Models/DisneyResort.swift
import Foundation

/// All six Disney resort destinations worldwide.
enum DisneyResort: String, Codable, CaseIterable, Identifiable {
    case waltDisneyWorld
    case disneylandResort
    case tokyoDisneyResort
    case disneylandParis
    case hongKongDisneyland
    case shanghaiDisneyland

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .waltDisneyWorld:    return "Walt Disney World"
        case .disneylandResort:   return "Disneyland Resort"
        case .tokyoDisneyResort:  return "Tokyo Disney Resort"
        case .disneylandParis:    return "Disneyland Paris"
        case .hongKongDisneyland: return "Hong Kong Disneyland"
        case .shanghaiDisneyland: return "Shanghai Disneyland"
        }
    }

    var shortName: String {
        switch self {
        case .waltDisneyWorld:    return "WDW"
        case .disneylandResort:   return "DLR"
        case .tokyoDisneyResort:  return "TDR"
        case .disneylandParis:    return "DLP"
        case .hongKongDisneyland: return "HKDL"
        case .shanghaiDisneyland: return "SHDL"
        }
    }

    var location: String {
        switch self {
        case .waltDisneyWorld:    return "Orlando, Florida, USA"
        case .disneylandResort:   return "Anaheim, California, USA"
        case .tokyoDisneyResort:  return "Urayasu, Chiba, Japan"
        case .disneylandParis:    return "Chessy, Île-de-France, France"
        case .hongKongDisneyland: return "Lantau Island, Hong Kong"
        case .shanghaiDisneyland: return "Pudong, Shanghai, China"
        }
    }

    var parks: [DisneyPark] {
        switch self {
        case .waltDisneyWorld:
            return [.magicKingdom, .epcot, .hollywoodStudios, .animalKingdom]
        case .disneylandResort:
            return [.disneyland, .californiaAdventure]
        case .tokyoDisneyResort:
            return [.tokyoDisneyland, .tokyoDisneySea]
        case .disneylandParis:
            return [.disneylandParkParis, .waltDisneyStudiosPark]
        case .hongKongDisneyland:
            return [.hongKongDisneylandPark]
        case .shanghaiDisneyland:
            return [.shanghaiDisneylandPark]
        }
    }

    /// The "flagship" park used for theming when no specific park is selected.
    var primaryPark: DisneyPark {
        parks.first ?? .magicKingdom
    }

    var castleAssetName: String {
        switch self {
        case .waltDisneyWorld:    return "castle-cinderella"
        case .disneylandResort:   return "castle-sleeping-beauty"
        case .tokyoDisneyResort:  return "castle-tokyo"
        case .disneylandParis:    return "castle-paris"
        case .hongKongDisneyland: return "castle-magical-dreams"
        case .shanghaiDisneyland: return "castle-enchanted-storybook"
        }
    }
}
